### ISSUE TITLE
webpack-extraction-plugin: support for code split CSS

### ISSUE THREAD

**Issue by layershifter** [microsoft/griffel#419]
Currently CSS extraction forces all styles to be inside one style sheet (one file). That simplifies the implementation, but we want to be able to only load the CSS that is used in loaded modules instead of loading whole CSS from a project.

There main constraint is that all stylesheets need to be sorted in the same way as in a single file (https://griffel.js.org/react/guides/atomic-css#lvfha-order-of-pseudo-classes) i.e. follow the approach with style buckets.

This means that only CSS from `d` bucket is safe, all other buckets are considered unsafe.

### Why is this hoisting required?

If the order is not consistent ➡️ applied CSS won't be consistent and results for different experiences are dependant on what chunks were loaded and when. Here is an example:

```css
/** css-1.css **/

.baz { color: "green" }
.baz:active { color: "magenta" }
```

```css
/** css-2.css **/
.baz:hover { color: "pink" }
```

```html
<link rel="./css-1.css" />
<link rel="./css-2.css" />
```

In this example the resulting order will be:

```css
.baz { color: "green" }
.baz:active { color: "magenta" }
.baz:hover { color: "pink" } /** 💣 :hover is later than :active **/
```

While it **should** be:

```css
.baz { color: "green" }
.baz:hover { color: "pink" }
.baz:active { color: "magenta" } /** ✅ :active is after :hover **/
```

**Comment by layershifter** [comment]
TL;DR we decided to stop working on it and keep it as it is for now.
 
---
 
## Problem
 
To support CSS chunks with Griffel we need to keep the order of CSS rules (https://griffel.js.org/react/guides/atomic-css#lvfha-order-of-pseudo-classes). With a single file as we have now - it's easy, we just sort CSS rules inside a file:
 
```css
/* griffel.css */
.foo {}
.bar {}
 
@media { .foo {} }
```

And this works fine 👍 

With multiple CSS files we need to keep order between files:

```css 
/* THIS IS ALREADY BROKEN 💣 */

/* griffel-main.css */
.foo { color: red }
@media { .foo { border-left-color: magenta } }
 
/* griffel-appbar.css */
.bar { border-left-color: blue }
```
 
`@media` and other at-rules don't have specificity. Once, `.foo` and `.bar` are applied to the same element - we screwed up:

```html
<div class="foo bar"></div>
```

There ⬆️  `.bar` wins over `.foo` in a media query. It's a regression compared to a previous scenario with a single file 💥 
 
How to deal with it? Options?
 
## Approach A: "simple"
 
_It's based on a POC in CompiledCSS._

We have rules without specificity i.e. `.foo {}`, we don't need to order them in a special way. For example, in the previous scenario we can do following:
 
```css
/* griffel-appbar.css */
.bar { border-left-color: blue }
 
/* griffel-main.css */
.foo { color: red }
 
/* griffel-unsafe-rules.css */
@media { .foo { border-left-color: magenta } }
 ```

- Order of `griffel-appbar.css` & `griffel-main.css` does not matter.
- Rules that can cause problems are under `griffel-unsafe-rules.css`.
 
Sounds simple? It was before we introduced `makeResetStyles()` 🙃

Rules from `makeResetStyles()` should be before atomic classes, so we should have following:
 
```css
/* griffel-reset-rules.css */
.reset {}
 
/* griffel-appbar.css */
.bar { border-left-color: blue }
 
/* griffel-main.css */
.foo { color: red }
 
/* griffel-unsafe-rules.css */
@media { .foo { border-left-color: magenta } }
```
 
The worst part is that `griffel-reset-rules.css`, `griffel-main.css` & `griffel-unsafe-rules.css` are present initially i.e. are the part of the initial chunk, but `griffel-appbar.css`, `griffel-calling-app.css`, etc. will be inserted in runtime by `css-loader`.
 
Result is that it becomes extremely challenging to insert these in a proper order.
 
Pros:
- it _works_

Cons:
- it's extremely complex
- it will require some insane tricks and probably will not be usable without a special config

## Approach B: use \@layer
 
Something new from CSS 🎉  Layers allow to sort rules based a custom order, here is the example:

```css 
/* griffel.css *
@layer a, b;
 
@layer b { .foo {} }
@layer a { .bar {} }
```
 
Looks cool, we can use layers in the same way style buckets and do sorting across files 🔥 **But...**
 
Layers ignore specificity outside their scope:

```css
/* griffel.css */
@layer a, b;
 
@layer b { .foo {} }
@layer a { .bar.foo {} }
```

While `.bar.foo` is more specific, `.foo` still wins ([example](https://codesandbox.io/s/determined-leaf-p6p7l2)). Fair enough, it's by the spec 🥹
 
So, while, in general, this approach _works_ and is **extremely** efficient for chunking, we will have some nasty visual regressions. The bigger issue that it's not easy to catch them and might be hard to fix them.
 
Pros:
- It's extremely efficient

Cons:
- Visual regressions
- Browser support (Chrome/Edge 99+)

## Recals problem
 
Another issue is that we get additional style recalculations. _Not so bad as with runtime insertion... Well probably not so bad._
 
Chunks will have duplicated CSS rules by design:

```css
/* griffel-appbar.css */
.foo {}
.bar {}
 
/* griffel-chat.css */
.foo {}
.baz {}
```
 
But, if there are 1k elements on a page with `.foo` class ➡️  we will get a nasty relalcs like this:
 
![image](https://github.com/microsoft/griffel/assets/14183168/42583c8b-a80c-4780-a9ed-0b6b8e640afa)

With Atomic CSS we have high risks for this event with almost every chunk loaded 🚨

IMO we don't think that we really want this. It also seems that Meta came to the same conclusion with Stylex:
 
> Most styling solutions lead to CSS that grows linearly with the size of your codebase. For larger applications it makes loading all your CSS upfront infeasible. So you end up splitting your CSS into multiple files and loading them lazily as you need them.
> 
> This improves your initial load time and performance, but it comes at the cost of slower updates as the browser needs to recalculate styles every time more CSS is loaded.
> 
> https://stylex-docusaurus.vercel.app/docs/design/principles/#a-single-file-is-better-than-lots-of-lazily-loaded-css
